### PR TITLE
Resizing buffer based on the size of its content

### PR DESCRIPTION
**Pull request by abdelq**
Since the buffer is based on the arc radius and thickness, it is often small enough so that keymap names like "French (Canada)" or "English (the divide/multiply keys toggle the layout)" are truncated.

This is more pronounced when a bigger font size is used.

The simplest solution is pretty much to use a fullscreen buffer.

Related to #88

### CODE REVIEW

**Comment by emersion** [comment]
>The simplest solution is pretty much to use a fullscreen buffer.

But then this damages the whole screen when typing. Instead, I'd prefer to resize the subsurface.

**Comment by abdelq** [comment]
@emersion I agree, but how would you get the `fe.height` and `extents.width` needed to calculate the proper subsurface when `cairo_create` is called on a `get_next_buffer`?

**Comment by emersion** [comment]
The way we've done it elsewhere is by rendering first to know the size, then rendering again if the size has changed:

```c
int previous_width; // width we used for the last frame
struct buffer *buffer = get_next_buffer(previous_width);
int new_width = render(previous_buffer);
if (new_width != previous_width) {
    // size changed, get a buffer with new width
    buffer = get_next_buffer(new_width);
    render(buffer);
}
// submit buffer as usual
```

It may be necessary to update `pool-buffer.c` to be able to create dummy 0-sized buffers for the very first frame. See e.g. https://github.com/emersion/mako/blob/master/pool-buffer.c#L125.

Let me know if something's unclear.

File: render.c [hunk 1]
```diff
@@ -265,7 +265,6 @@ void render_frame(struct swaylock_surface *surface) {
 			// border
 			cairo_set_source_u32(cairo, state->args.colors.layout_border);
 			cairo_stroke(cairo);
-			cairo_new_sub_path(cairo);
```
Comment (abdelq) [inline render.c]: See [documentation](https://www.cairographics.org/manual/cairo-Paths.html#cairo-new-sub-path) for rationale

**Comment by abdelq** [comment]
@emersion Updated the PR.

File: password.c [hunk 2]
```diff
@@ -132,6 +132,8 @@ void swaylock_handle_key(struct swaylock_state *state,
 	case XKB_KEY_Alt_R:
 	case XKB_KEY_Super_L:
 	case XKB_KEY_Super_R:
+	case XKB_KEY_ISO_Prev_Group:
```
Comment (emersion) [inline password.c]: This probably doesn't belong to this pull request.

File: render.c [hunk 3]
```diff
@@ -274,6 +280,19 @@ void render_frame(struct swaylock_surface *surface) {
...
+
+			new_height += fe.height + 2 * box_padding;
+			if (new_width < (extents.width + 2 * box_padding)) {
+				new_width = extents.width + 2 * box_padding;
+			}
+		}
+
+		if (buffer_width != new_width || buffer_height != new_height) {
+			destroy_buffer(surface->current_buffer);
+			surface->current_buffer->width = new_width;
```
Comment (emersion) [inline render.c]: Hijacking `current_buffer` to store the current width/height is a little strange to me. Maybe we can add some `indicator_{width,height}` fields instead?

File: password.c (hunk 2 above)
Comment (abdelq) [reply password.c]: Well, it's certainly more related than [3db0b6](https://github.com/swaywm/swaylock/pull/89/commits/3db0b6). The point of resizing the buffer is to properly show the keymap name, which is usually changed with those keys.

Comment (emersion) [reply password.c]: When the keymap changes, a `wl_keyboard.modifier` event is triggered. This is where we should re-render, not here.

I see that you've split this PR into two commits, which seems like a good idea to me.

File: render.c [hunk 4]
```diff
@@ -274,6 +280,18 @@ void render_frame(struct swaylock_surface *surface) {
 			cairo_set_source_u32(cairo, state->args.colors.layout_text);
 			cairo_show_text(cairo, layout_text);
 			cairo_new_sub_path(cairo);
+
+			new_height += fe.height + 2 * box_padding;
+			if (new_width < (extents.width + 2 * box_padding)) {
```
Comment (emersion) [inline render.c]: Style: unnecessary parentheses

File: render.c [hunk 5]
```diff
@@ -274,6 +280,18 @@ void render_frame(struct swaylock_surface *surface) {
 			cairo_set_source_u32(cairo, state->args.colors.layout_text);
 			cairo_show_text(cairo, layout_text);
 			cairo_new_sub_path(cairo);
+
+			new_height += fe.height + 2 * box_padding;
```
Comment (emersion) [inline render.c]: Hmm, is this correct? The buffer will grow vertically each time we render.

(IMHO we should probably set `new_width = new_height = 0` at the beginning of the function, always computing the new size without taking into account the old one)

File: password.c (hunk 2 above)
Comment (abdelq) [reply password.c]: Doesn't seem to be the case. You can try it out with and without that commit and see for yourself.

File: render.c (hunk 5 above)
Comment (abdelq) [reply render.c]: Yes this is correct. The compute of the new size doesn't take into account the old. Actually, its starting value is `buffer_diameter`. If the keymap indicator is to be rendered too, we need to take it into account and add it to the height.

File: password.c (hunk 2 above)
Comment (emersion) [reply password.c]: I do get a `wl_keyboard.modifiers` event when switching layouts. Instead of handling these keys, please damage the state in `keyboard_modifiers`.

File: render.c (hunk 5 above)
Comment (emersion) [reply render.c]: Oh, indeed, I got confused with `buffer_width/height` vs. `buffer_diameter`

File: password.c (hunk 2 above)
Comment (abdelq) [reply password.c]: Updated. @emersion, thank you for your patience.
